File: frontend/src/types/Reports.ts
import type { ExecutionStatus, ProjectStage } from './common';
import type { DesignerWorkload } from './Dashboard';

export interface ProjectHoursReportRow {
  project_id: string;
  tool_number: string;
  part_description: string;
  customer_name: string;
  quoted_hours: number;
  actual_hours: number;
  hours_variance: number;
  execution_status: ExecutionStatus;
  project_stage: ProjectStage;
  contributors?: import('./TimesheetEntry').ProjectContributorSummary[];
}

export interface CustomerSummaryReportRow {
  customer_id: string;
  customer_name: string;
  project_count: number;
  total_quoted_hours: number;
  total_actual_hours: number;
  hours_variance: number;
}

export interface ProductiveHoursReportRow {
  project_id: string | null;
  tool_number: string | null;
  customer_name: string | null;
  task_type_name: string | null;
  total_hours: number;
  billable_hours: number;
  non_billable_hours: number;
}

export interface NonProductiveHoursReportRow {
  non_productive_code: string;
  description: string;
  customer_name: string | null;
  total_hours: number;
}

export interface BillableUtilizationReportRow {
  user_id: string;
  designer_name: string;
  billable_hours: number;
  non_billable_hours: number;
  np_hours: number;
  billable_percent: number;
  non_billable_percent: number;
}

export interface MonthlyNpTrendReportRow {
  month: string;
  total_np_hours: number;
}

export interface NpHoursByDesignerReportRow {
  user_id: string;
  designer_name: string;
  total_np_hours: number;
}

export interface BillableVsNonBillableReportRow {
  billable_hours: number;
  non_billable_hours: number;
  np_hours: number;
  leave_days: number;
  billable_percent: number;
  non_billable_percent: number;
}

export interface TopNpActivityReportRow {
  non_productive_code: string;
  description: string;
  total_hours: number;
  entry_count: number;
}

export interface ProjectPortfolioReportRow {
  project_id: string;
  tool_number: string;
  customer_name: string;
  project_stage: ProjectStage;
  execution_status: ExecutionStatus;
  due_date: string;
  health: string;
}

export interface ProjectStageSummaryRow {
  project_stage: ProjectStage;
  project_count: number;
}

export interface ExecutionStatusSummaryRow {
  execution_status: ExecutionStatus;
  project_count: number;
}

export interface TimesheetExportReportRow {
  entry_date: string;
  employee_name: string;
  team_name?: string | null;
  customer_name?: string | null;
  tool_number?: string | null;
  task_name?: string | null;
  hours: number;
  is_billable: boolean;
  work_category: string;
}

export type { DesignerWorkload };

export interface ProjectsByTeamReportRow {
  team_id: string;
  team_name: string;
  team_colour: string;
  project_count: number;
}

export interface HoursByTeamReportRow {
  team_id: string;
  team_name: string;
  quoted_hours: number;
  actual_hours: number;
  hours_variance: number;
}

export interface QuotedVsActualByTeamReportRow {
  team_id: string;
  team_name: string;
  quoted_hours: number;
  actual_hours: number;
  variance_hours: number;
  variance_percent: number;
}

export interface TeamUtilizationReportRow {
  team_id: string | null;
  team_name: string;
  member_count: number;
  allocated_hours: number;
  actual_hours: number;
  utilization_percent: number;
}

export interface CustomerByTeamReportRow {
  team_id: string;
  team_name: string;
  customer_id: string;
  customer_name: string;
  project_count: number;
}

export interface DesignerByTeamReportRow {
  team_id: string;
  team_name: string;
  user_id: string;
  user_name: string;
  role_within_team: string | null;
}

export interface TeamProfitabilityReportRow {
  team_id: string;
  team_name: string;
  quoted_hours: number;
  actual_hours: number;
  margin_hours: number;
  margin_percent: number;
}

export interface MonthlyTeamSummaryRow {
  team_id: string;
  team_name: string;
  year: number;
  month: number;
  actual_hours: number;
}

File: app/schemas/reports.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel

from app.models.enums import ExecutionStatus, MilestoneStatus, ProjectHealth, ProjectStage, TimesheetStatus, WorkCategory
from app.schemas.dashboard import DesignerWorkload
from app.schemas.timesheet import ProjectContributorSummary


class ProjectHoursReportRow(BaseModel):
    project_id: UUID
    tool_number: str
    part_description: str
    customer_name: str
    quoted_hours: Decimal
    actual_hours: Decimal
    hours_variance: Decimal
    execution_status: ExecutionStatus
    project_stage: ProjectStage
    contributors: list[ProjectContributorSummary] = []


class CustomerSummaryReportRow(BaseModel):
    customer_id: UUID
    customer_name: str
    project_count: int
    total_quoted_hours: Decimal
    total_actual_hours: Decimal
    hours_variance: Decimal
    designers_used: int = 0
    teams_used: int = 0


class ProjectsByTeamReportRow(BaseModel):
    team_id: UUID
    team_name: str
    team_colour: str
    project_count: int


class HoursByTeamReportRow(BaseModel):
    team_id: UUID
    team_name: str
    quoted_hours: Decimal
    actual_hours: Decimal
    hours_variance: Decimal


class QuotedVsActualByTeamReportRow(BaseModel):
    team_id: UUID
    team_name: str
    quoted_hours: Decimal
    actual_hours: Decimal
    variance_hours: Decimal
    variance_percent: Decimal


class TeamUtilizationReportRow(BaseModel):
    team_id: UUID | None = None
    team_name: str
    member_count: int
    allocated_hours: Decimal
    actual_hours: Decimal
    utilization_percent: Decimal


class CustomerByTeamReportRow(BaseModel):
    team_id: UUID
    team_name: str
    customer_id: UUID
    customer_name: str
    project_count: int


class DesignerByTeamReportRow(BaseModel):
    team_id: UUID
    team_name: str
    user_id: UUID
    user_name: str
    role_within_team: str | None = None


class TeamProfitabilityReportRow(BaseModel):
    team_id: UUID
    team_name: str
    quoted_hours: Decimal
    actual_hours: Decimal
    margin_hours: Decimal
    margin_percent: Decimal


class MonthlyTeamSummaryRow(BaseModel):
    team_id: UUID
    team_name: str
    year: int
    month: int
    actual_hours: Decimal


class TeamResourcePlanningRow(BaseModel):
    team_id: UUID
    team_name: str
    team_colour: str
    member_count: int
    capacity_hours: Decimal
    allocated_hours: Decimal
    actual_hours: Decimal
    remaining_capacity_hours: Decimal
    utilization_percent: Decimal


class ReportsBundle(BaseModel):
    project_hours: list[ProjectHoursReportRow]
    designer_utilization: list[DesignerWorkload]
    customer_summary: list[CustomerSummaryReportRow]


class TimesheetApprovalReportRow(BaseModel):
    timesheet_id: UUID
    user_name: str
    week_start: date
    status: TimesheetStatus
    total_hours: Decimal
    approved_by_name: str | None = None
    approval_comments: str | None = None


class ProjectDelayReportRow(BaseModel):
    project_id: UUID
    tool_number: str
    customer_name: str
    due_date: date
    days_overdue: int
    health: ProjectHealth
    execution_status: ExecutionStatus
    project_stage: ProjectStage


class MilestoneCompletionReportRow(BaseModel):
    project_code: str
    milestone_name: str
    status: MilestoneStatus
    due_date: date | None = None
    completed_at: datetime | None = None


class DesignerProductivityReportRow(BaseModel):
    user_id: UUID
    designer_name: str
    role: str
    approved_hours: Decimal
    submitted_hours: Decimal
    draft_hours: Decimal


class ProductiveHoursReportRow(BaseModel):
    project_id: UUID | None = None
    tool_number: str | None = None
    customer_name: str | None = None
    task_type_name: str | None = None
    total_hours: Decimal
    billable_hours: Decimal
    non_billable_hours: Decimal


class NonProductiveHoursReportRow(BaseModel):
    non_productive_code: str
    description: str
    customer_name: str | None = None
    total_hours: Decimal


class BillableUtilizationReportRow(BaseModel):
    user_id: UUID
    designer_name: str
    billable_hours: Decimal
    non_billable_hours: Decimal
    np_hours: Decimal
    billable_percent: Decimal
    non_billable_percent: Decimal


class MonthlyNpTrendReportRow(BaseModel):
    month: str
    total_np_hours: Decimal


class NpHoursByDesignerReportRow(BaseModel):
    user_id: UUID
    designer_name: str
    total_np_hours: Decimal


class BillableVsNonBillableReportRow(BaseModel):
    billable_hours: Decimal
    non_billable_hours: Decimal
    np_hours: Decimal
    leave_days: int = 0
    billable_percent: Decimal
    non_billable_percent: Decimal


class TopNpActivityReportRow(BaseModel):
    non_productive_code: str
    description: str
    total_hours: Decimal
    entry_count: int


class ProjectPortfolioReportRow(BaseModel):
    project_id: UUID
    tool_number: str
    customer_name: str
    project_stage: ProjectStage
    execution_status: ExecutionStatus
    due_date: date
    health: ProjectHealth


class ProjectStageSummaryRow(BaseModel):
    project_stage: ProjectStage
    project_count: int


class ExecutionStatusSummaryRow(BaseModel):
    execution_status: ExecutionStatus
    project_count: int


class TimesheetExportReportRow(BaseModel):
    entry_date: date
    employee_name: str
    team_name: str | None = None
    customer_name: str | None = None
    tool_number: str | None = None
    task_name: str | None = None
    hours: Decimal
    is_billable: bool
    work_category: WorkCategory
